### PR TITLE
RequireJS optimization

### PR DESCRIPTION
**Pull request by smagch**
I think libraries like `jquery`, `underscore` and  `backbone` would be maintainable with `require dependencies` though it makes config relatively large.

I have no idea whether thimble do conditional loading or not.
If it don't do dynamic conditional loading, require.js would be replaced with [almond.js](https://github.com/jrburke/almond).

This pull request have some points which should be improve: duplicate config in `index.html`, `test/index.html`, `build-require.js`... :(
I'm making config a JSON file and inserting it during server-side templating in my personal backbone project.
I haven't looked into webpagemaker, so you may know better solutions.

toolness/friendlycode#112

### CODE REVIEW

**Comment by toolness** [comment]
Very cool, thanks for all this work @smagch! Friendlycode doesn't currently do conditional loading, with the possible exception of TypeKit, so RequireJS should _probably_ be fine. I'll take a look at your stuff today, sorry for the really long delay on this.

**Comment by smagch** [comment]
> There isn't a Python variant of this

`r.js` also [works with Rhino](https://github.com/jrburke/r.js/#java) though I haven't tried that: I haven't set up `.jar` file by myself because Eclipse do everything when I was working as a Java developer.

It would be nice not to merge build script I wrote because it has ugly copy and paste of config file :(

I'm using Makefile and node build script to build my app which use `require.js` and `backbone.js` in my personal project.
https://github.com/dogazei/dogazei.org/blob/v0.1.2/Makefile#L67-83
https://github.com/dogazei/dogazei.org/blob/v0.1.2/support/buildRequire.js
During development, I just simply comment out optimized script and use `main.js`
https://github.com/dogazei/dogazei.org/blob/v0.1.2/views/holy-potato/index.jade#L18-22
Before deployment, I build scripts with Makefile and comment out `main.js` and use optimized script.

When you install r.js via npm, it seems to be able to optimize by command line! But I haven't figure out how to use that correctly: I ended up with error.

**Comment by toolness** [comment]
I've just offered a few fixes to this here: https://github.com/smagch/friendlycode/pull/1. Let me know what you think.

**Comment by toolness** [comment]
Hmm, found one bug w/ the pull request, I think, which is that error messages don't seem to work anymore. If I create an invalid tag like `<pe>`, for instance, no error message appears and this error gets logged to the console:

```
Error template not found for INVALID_TAG_NAME @ http://localhost:8001/slowparse/spec/errors.jquery.js:91
```

The error templates are supposed to be loaded as part of the `appReady` requirejs plugin, I think.

**Comment by smagch** [comment]
Ah, I didn't notice the bug! It's weird... help messages are shown correctly.

File: js/template.js [hunk 1]
```diff
@@ -1,8 +1,32 @@
...
+        }
+
+        onLoad(template);
+      });
+    },
+    write: function (pluginName, moduleName, write) {
+      if (buildMap[moduleName]) {
+        var content = buildMap[moduleName];
+        write.asModule(pluginName + "!" + moduleName,
+          "define(function () { \n  return " + content + ";});\n");
```
Comment (smagch) [inline js/template.js]: I missed dependencies. It need to be

``` javascript
"define(['underscore'], function (_) { \n  return " + content + ";});\n");
```

**Comment by toolness** [comment]
Hmm, something else strange seems to be going on--it's as though two copies of `jQuery` are in the page at once. I'm getting an error that `$.fn.fillError()` doesn't exist in `fc/ui/error-help.js`, even though the file it's defined in (`slowparse/spec/errors.jquery.js`) has been loaded beforehand.

**Comment by smagch** [comment]
> (`slowparse/spec/errors.jquery.js` ) has been loaded beforehand.

[error-help.js](https://github.com/toolness/friendlycode/blob/smagch-optimize/js/fc/ui/error-help.js#L5) currently don't have `jquery-slowparse` dependencies. So it could be possible to error, couldn't it?

File: js/template.js (hunk 1 above)
Comment (smagch) [inline js/template.js]: 49a0d34

**Comment by toolness** [comment]
> [error-help.js](https://github.com/toolness/friendlycode/blob/smagch-optimize/js/fc/ui/error-help.js#L5) currently don't have `jquery-slowparse` dependencies. So it could be possible to error, couldn't it?

Well, the exception coming out of error-help is raised a long time after initial page load, when I modify a document to contain an error... I think the missed dependency would be a good explanation if the exception were happening during load time, but since the dependency is just a shim, the jquery-slowparse plugin should have modified the global `jQuery` object by the time I've typed an error into the editor.

**Comment by toolness** [comment]
Huh, something you did fixed the problem. Everything looks good now, so I'm going to merge this pull request. Thanks for everything!

**Comment by toolness** [comment]
It looks like the multiple-instances-of-jQuery problem I was observing was due to tabzilla trying to load jQuery itself! Fixed this in dfbd4c21335868f3f1e30f8a35852af6e248107a.

**Comment by smagch** [comment]
I could add `include: "require.min"` in [build config](https://github.com/toolness/friendlycode/blob/dfbd4c21335868f3f1e30f8a35852af6e248107a/build-require.js#L21-31).
It'll concatenate `require.js` and optimized main script.

Instead of 

``` html
<script src="js/require.min.js" data-main="js/main-built.js"></script>
```

It'll be simply like this.

``` html
<script src="js/main-built.js"></script>
```

I just downloaded [almond.js](https://github.com/jrburke/almond) and tried `include: "almond"`. But it didn't work out: it seems `require("typekit-ready!")` have a trouble.
Anyway, I'll let you know if I have time and figure out a reason of the error.

Thanks, I've learned a lot through the pull request.

**Pull request merged**